src/cryptonote_basic/cryptonote_format_utils.cpp

```diff
@@ -992,20 +992,39 @@ namespace cryptonote
     block_hashes_cached = block_hashes_cached_count;
   }
   //---------------------------------------------------------------
-  crypto::secret_key encrypt_key(const crypto::secret_key &key, const std::string &passphrase)
+  crypto::secret_key encrypt_key(crypto::secret_key key, const std::string &passphrase)
   {
     crypto::hash hash;
     crypto::cn_slow_hash(passphrase.data(), passphrase.size(), hash);
     sc_add((unsigned char*)key.data, (const unsigned char*)key.data, (const unsigned char*)hash.data);
     return key;
   }
   //---------------------------------------------------------------
-  crypto::secret_key decrypt_key(const crypto::secret_key &key, const std::string &passphrase)
+  crypto::secret_key decrypt_key(crypto::secret_key key, const std::string &passphrase)
   {
     crypto::hash hash;
     crypto::cn_slow_hash(passphrase.data(), passphrase.size(), hash);
     sc_sub((unsigned char*)key.data, (const unsigned char*)key.data, (const unsigned char*)hash.data);
     return key;
   }
-
+  //---------------------------------------------------------------
+  std::string encrypt_data_deterministic(const std::string &plaintext, const std::string &passphrase)
+  {
+    crypto::chacha8_iv iv;
+    memset(&iv, 0, sizeof(iv));
+    std::string ciphertext;
+    ciphertext.resize(plaintext.size());
+    static_assert(HASH_SIZE == sizeof(crypto::hash), "Bad hash size");
+    static_assert(HASH_SIZE == CHACHA8_KEY_SIZE, "chacha8 key size does not match hash size");
+    unsigned char key[HASH_SIZE];
+    crypto::cn_slow_hash(passphrase.data(), passphrase.size(), (char*)key);
+    sc_reduce32(key);
+    crypto::chacha8(&plaintext[0], plaintext.size(), key, iv.data, &ciphertext[0]);
```

Comment:
This uses the same key/iv pair for all `plaintext` for a specific `passphrase`. This breaks confidentiality if multiple messages are encrypted. The iv in this context is actiually a nonce - it is to be used only once. Creating a deterministic nonce is a tricky because the nonce will have to be stored with the encrypted data for decryption, which could unintentionally leak the plaintext contents. The best construction I can up with right now is `hash(plaintext | hash(key))` where `|` is concatenation (this is definitely not standard usage). Also, there is no MAC so encryption manipulation will not be detected.

Reply:
Is tamper detection important here? This is one of the issues I'm having to refit in LMDB now, tweaking the crypto handlers to work with authenticated encryption algorithms. (Which requires reserving additional space for the MAC/signature/whatever.)

Reply:
Why would the ciphertext always need to be the same? 

Also, would this be used to encrypt the language (english, etc) seed words OR a binary version of the seed? Recovering useful information from random binary data is unrealistic - although I imagine most crypto experts still might be horrified. If this is encrypting a language seed then reusing the iv/key is risky since it will leak biases in the language.

The other concern would be someone using this for an unintended purpose in the future. If the key/iv can be reused, a comment to caution usage would be helpful.

Reply:
@hyc thats a good question. Is the use case -> encrypt binary scalar with passphrase, then give user language equivalent? That way if someone finds the recovery words, the passphrase would also have to be known? In that context, the MAC will tell you immediately if transcription is correct. The user would most likely find out "soon enough" if it was wrong anyway, and I cannot think of a scenario where an attacker would want to manipulate this a meaningful way (other than to be a jerk and cause you to lose XMR). I mentioned it previously because I thought this was going to be for general purpose usage.

Reply:
I want the ciphertext to be the same because it's rendered as words visible to the user, so they'll get mighty confused if they get different seeds. It doens't *need* to be the same technically. Just a user confusion thing.
The seed does not include language, etc. Just the necessary keys (plus M/N numbers).
The seed, even encrypted, is supposed to be kept by the user, and not sent to anyone. The encryption is a secondary defense. I expect the password to be weak. I really do not expect any tampering by an adversary in this case. Partial destruction is possible though.

Reply:
What are M/N numbers? What type of data is it? Make sure that the resulting ciphertext is _always_ the same size, as that could be another source of unintentional information leakage.

I would also leave a comment in the header on how this is implemented - particular that there is no MAC. This is not a generic encryption utility, so its worth cautioning maintainers to review any code that calls it.

Reply:
They are the threshold and total signers in multisig, uint32_t both. The plaintext size will leak the type of multisig M/N. Encryption typically does not imply authenticated encryption anyway, does it ?